standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+---
+title: WAKU2-RLN-CONTRACT
+name: Waku2 RLN Contract Specification
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This document describes membership management for the RLN smart contract, in particular:
+- membership-related contract functionality;
+- suggested parameters valued for the initial mainnet deployment;
+- contract governance and upgradability.
+
+This document currently only considers membership-related functionality.
+It might later evolve into a full-fledged contract specification.
+
+RLN is only deployed on Sepolia testnet ([source code](https://github.com/waku-org/waku-rlnv2-contract/blob/main/src/WakuRlnV2.sol)) as of August 2024.
+This document aims to outline the path to its mainnet deployment.
+
+## Background
+
+Rate-Limiting Nullifier (RLN) is a ZK-based gadget used for privacy-preserving rate limiting in Waku.
+The RLN smart contract is the core element of RLN architecture.
+The smart contract stores the RLN tree that contains all currently existing memberships.
+Users interact with the contract to manage their memberships
+and to get the necessary data for proof generation and verification.
+
+Sending messages is handled by Waku Relay nodes.
+To send a message, the sender MUST attach a ZK-proof to the message.
+A Relay node MUST relay a message unless:
+- the message belongs to any epoch other than the current one; or
+- the user has exceed their allowed rate limit for the current epoch; or
+- the ZK-proof fails to prove that the sender holds a membership.
+
+The full specification of Relay is out of scope for this document.
+
+## Contract overview
+
+The contract MUST provide the following functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+Contract parameters and their RECOMMENDED values for the initial mainnet deployment are as follows:
+
+| Parameter                                   | Symbol    | Value   | Units                              |
+| ------------------------------------------- | --------- | ------- | ---------------------------------- |
+| Epoch length                                | `epoch`   | `10`    | minutes                            |
+| Maximum total rate limit of all memberships | `R_{max}` | `20000` | messages per `epoch`               |
+| Minimal rate limit of one membership        | `r_{min}` | `20`    | messages per `epoch`               |
+| Price of `1` message per epoch              | `p_u`     | `0.01`  | `USD` per one period of length `T` |
+| Membership expiration term                  | `T`       | `90`    | days                               |
```

Comment:
Intuition here is that `90` days may be too short. Wouldn't year long memberships be more useful? Requiring user action every 3 months would be a frequent burden on the app or end users.

Reply:
As far as I recall (I couldn't find the comment quickly), initially we discussed 1 year, then @fryorcraken suggested 3 or 6 months, and I picked the former. The rationale was that during initial development stages we don't want users to be committed to the contract as it existed at their registration time, as the _Owner_ (i.e., we) might change parameter values relatively frequently.

Do you think 6 months could be a reasonable middle ground? Plus 1 month of grace period.

Reply:
No strong opinions, but I'd indeed feel more comfortable with 6 months.

Reply:
6 months sound good

Reply:
Great, let's make it 6 months (180 days, defining a month as 30 days).